src/batch/src/executor/row_seq_scan.rs

```diff
@@ -93,20 +95,47 @@ fn is_full_range<T>(bounds: &impl RangeBounds<T>) -> bool {
         && matches!(bounds.end_bound(), Bound::Unbounded)
 }
 
-fn get_scan_bound(scan_range: ScanRange) -> (Row, impl RangeBounds<Datum>) {
+fn cast(lit: LiteralExpression, return_ty: DataType) -> Datum {
+    let (data, data_ty) = (lit.literal(), lit.return_type());
+    if data_ty == return_ty {
+        return data;
+    }
+
+    let data_chunk = DataChunk::from_rows(&[Row(vec![data])], &[data_ty.clone()]).unwrap();
+    let expr = ExprNode {
+        expr_type: Type::Cast as i32,
+        return_type: Some(return_ty.to_protobuf()),
+        rex_node: Some(RexNode::FuncCall(FunctionCall {
+            children: vec![make_input_ref(0, data_ty.prost_type_name())],
+        })),
+    };
+    let vec_executor = build_from_prost(&expr).unwrap();
+    let array = vec_executor.eval(&data_chunk).unwrap();
+    array.iter().next().unwrap().to_owned_datum()
+}
+
+fn get_scan_bound(
+    scan_range: ScanRange,
+    mut pk_types: impl Iterator<Item = DataType>,
+) -> (Row, impl RangeBounds<Datum>) {
     let pk_prefix_value = Row(scan_range
         .eq_conds
         .iter()
-        .map(|v| LiteralExpression::try_from(v).unwrap().literal())
+        .map(|v| {
+            let lit = LiteralExpression::try_from(v).unwrap();
+            let ty = pk_types.next().unwrap();
+            cast(lit, ty)
+        })
         .collect_vec());
     if scan_range.lower_bound.is_none() && scan_range.upper_bound.is_none() {
         return (pk_prefix_value, (Bound::Unbounded, Bound::Unbounded));
     }
 
+    let bound_ty = pk_types.next().unwrap();
     let build_bound = |bound: &scan_range::Bound| -> Bound<Datum> {
-        let datum = LiteralExpression::try_from(bound.value.as_ref().unwrap())
-            .unwrap()
-            .literal();
+        let lit = LiteralExpression::try_from(bound.value.as_ref().unwrap()).unwrap();
+
+        let datum = cast(lit, bound_ty.clone());
```

Comment:
I think it might be better to store `Datum` with the correct data type in `ScanRange` instead of converting it here. 🤔

Reply:
I agree, but we don't have eval in front yet? 🤔

Reply:
> I agree, but we don't have eval in fe yet? 🤔

We need one!